File: src/photo.ts
import { boundary } from '@toba/tools';
import { geoJSON, IMappable } from '@toba/map';
import { PhotoSize, EXIF, PostProvider } from './index';
import { ensurePostProvider } from './providers';

export class Photo implements IMappable<GeoJSON.Point> {
   /** Provider photo ID */
   id: string;
   /** Position of photo within post. */
   index: number;
   sourceUrl: string;
   title: string;
   description?: string;
   /** Tags applied to the photo. */
   tags: Set<string> = new Set();
   dateTaken?: Date;
   latitude?: number;
   longitude?: number;
   /** Whether this is the post's main photo. */
   primary: boolean = false;
   size: { [key: string]: PhotoSize } = {};
   /** Size at which to preview the photo such as in search results. */
   preview: PhotoSize;
   /** Normal photo size shown within post. */
   normal: PhotoSize;
   /** Size shown when post photo is clicked for enlargmenet. */
   big: PhotoSize;

   private _exif: EXIF;

   /**
    * Whether taken date is an outlier compared to other photos in the same
    * post. Outliers may be removed from mini-maps so the maps aren't overly
    * zoomed-out to accomodate contextual photos taken days before or after
    * the main post.
    *
    * @see http://www.wikihow.com/Calculate-Outliers
    */
   outlierDate?: boolean;

   constructor(id: string, index: number) {
      this.id = id;
      this.index = index;
   }

   private get load(): PostProvider<any> {
      return ensurePostProvider();
   }

   /**
    * Comma-delimited list of all tags applied to the photo.
    */
   get tagList(): string {
      return Array.from(this.tags).join(',');
   }

   async EXIF(): Promise<EXIF> {
      if (this._exif === null) {
         this._exif = await this.load.exif(this.id);
      }
      return this._exif;
   }

   /**
    * GeoJSON point constructor expect requires populated coordinate array or
    * `null`.
    */
   private get coordinate(): [number, number] | null {
      return this.longitude !== undefined && this.latitude !== undefined
         ? [this.longitude, this.latitude]
         : null;
   }

   /**
    * Generate GeoJSON for photo feature.
    *
    * @param partKey Optional series part that photo post belongs to, used to
    * generate link from map info box back to post URL
    */
   geoJSON(partKey?: string): GeoJSON.Feature<GeoJSON.Point> {
      const properties: MapPhoto = { url: this.size.preview.url };

      if (partKey !== undefined) {
         // implies GeoJSON for single post
         properties.title = this.title;
         properties.partKey = partKey;
      }
      return {
         type: geoJSON.Type.Feature,
         properties,
         geometry: geoJSON.geometry(geoJSON.Type.Point, this.coordinate)
      } as GeoJSON.Feature<GeoJSON.Point>;
   }
}

/**
 * Simplistic outlier calculation identifies photos that are likely not part of
 * the main sequence.
 *
 * @see https://en.wikipedia.org/wiki/Outlier
 * @see http://www.wikihow.com/Calculate-Outliers
 */
export function identifyOutliers(photos: Photo[]) {
   const fence = boundary(
      photos
         .filter(p => p.dateTaken !== undefined)
         .map(p => p.dateTaken!.getTime())
   );

   if (fence !== null) {
      for (const p of photos) {
         if (p.dateTaken === undefined) {
            continue;
         }
         const d = p.dateTaken.getTime();
         if (d > fence.max || d < fence.min) {
            p.outlierDate = true;
         }
      }
   }
}

/**
 * GeoJSON properties for photos.
 */
export interface MapPhoto {
   url?: string;
   title?: string;
   partKey?: string;
   /** Distance from clicked cluster */
   distance?: number;
}

File: src/photo.rs
use crate::{min_date, replace_pairs, ExifConfig};
use chrono::{DateTime, Local};

/// Latitude and longitude in degrees
#[derive(Debug, Default)]
pub struct Location {
    pub longitude: f64,
    pub latitude: f64,
}

impl Location {
    /// Whether latitude and longitude are within valid range
    pub fn is_valid(&self) -> bool {
        self.longitude <= 180.0
            && self.longitude >= -180.0
            && self.latitude <= 90.0
            && self.latitude >= -90.0
    }
}

impl PartialEq for Location {
    fn eq(&self, other: &Self) -> bool {
        self.longitude == other.longitude && self.latitude == other.latitude
    }
}

impl Eq for Location {}

#[derive(Debug)]
#[repr(u8)]
pub enum ExposureMode {
    Undefined = 0,
    Manual = 1,
    ProgramAE = 2,
    AperturePriority = 3,
    ShutterPriority = 4,
    Creative = 5,
    Action = 6,
    Portrait = 7,
    Landscape = 8,
    Bulb = 9,
}
impl Default for ExposureMode {
    fn default() -> Self {
        ExposureMode::Undefined
    }
}

/// Information about the camera used to make the photo.
// https://exiftool.org/TagNames/EXIF.html
#[derive(Debug, Default)]
pub struct Camera {
    /// Make and model of the camera
    pub name: String,
    /// Exposure compensation
    pub compensation: String,
    pub shutter_speed: String,
    pub mode: ExposureMode,
    pub aperture: String,
    pub focal_length: f64,
    pub iso: u32,
    pub lens: String,
}

#[derive(Debug)]
pub struct Photo {
    /// File name of the photo
    pub name: String,
    /// Name of photographer recorded in EXIF
    pub artist: String,
    /// Name of software used to process the photo
    pub software: String,
    pub title: String,
    pub caption: String,
    pub camera: Camera,
    pub location: Location,
    /// One-based position of photo within post
    pub index: u8,
    /// Tags applied to the photo
    pub tags: Vec<String>,
    /// Whether this is the post's main photo
    pub primary: bool,
    /// When the photograph was taken per camera EXIF
    pub date_taken: DateTime<Local>,

    /// Whether taken date is an outlier compared to other photos in the same
    /// post. Outliers may be removed from mini-maps so the maps aren't overly
    /// zoomed-out to accomodate contextual photos taken days before or after
    /// the main post.
    ///
    /// See http://www.wikihow.com/Calculate-Outliers
    pub outlier_date: bool,

    /// Whether values have been formatted based on configuration
    pub sanitized: bool,
}

impl Photo {
    pub fn sanitize(&mut self, config: &ExifConfig) {
        if self.sanitized {
            return;
        }
        self.software = replace_pairs(self.software.clone(), &config.software);
        self.camera.name =
            replace_pairs(self.camera.name.clone(), &config.camera);
        self.camera.lens =
            replace_pairs(self.camera.lens.clone(), &config.lens);

        self.sanitized = true;
    }
}

impl Default for Photo {
    fn default() -> Self {
        Photo {
            name: String::new(),
            artist: String::new(),
            software: String::new(),
            title: String::new(),
            caption: String::new(),
            camera: Camera::default(),
            location: Location::default(),
            index: 0,
            tags: Vec::new(),
            primary: false,
            date_taken: min_date(),
            outlier_date: false,
            sanitized: false,
        }
    }
}
